### DISCUSSION TITLE
Auto execution for graph

### DISCUSSION THREAD

**Discussion by Tokarzewski** [IndiePython/nodezator#26] [Feature Request]
Nodezator could automatically detect changes and recalculate changed nodes without the need to select F12.
This will be quite useful for running connected workflows that are fetching stream data from a database that is constantly changing its values.

**Comment by KennedyRichard** [comment]
This is interesting. I agree, this would be useful.

Just to ensure we are on the same page, you mean allowing the node graph to execute indefinitely based on a continuous data stream until manually or programmatically stopped, right?

**Reply by Tokarzewski** [reply]
Yes, indefinitely, until the data input is the same.

**Reply by Tokarzewski** [reply]
There could be two modes to run nodezator graphs, automatic and manual.

**Comment by KennedyRichard** [comment]
> Yes, indefinitely, until the data input is the same.

Neat, I'll add it to the list of requested features. But before that, I'd like to find a more precise name for the feature.

I think _auto execution_ or _automatic mode_ would be better suited for a mode that executes the graph every time the user changes an input whereas what you described is a mode that executes the graph according to data read from a stream.

The first alternative that come to mind is _daemon mode_. According to [wikipedia](https://en.wikipedia.org/wiki/Daemon_(computing)), daemons...

> ...runs as a background process, rather than being under the direct control of an interactive user. [...]
> [...]
> [...] will respond to network requests, hardware activity, or other programs by performing some task

Is this feature named _auto execution_ as well in other app you are familiar with?

Another alternatives that comes to mind are _event handler_ or _event listener mode_. I'll do a little research about those and other terms to make sure we pick the appropriate name in order for people to know exactly what to expect of the feature.

**Reply by nrichards** [reply]
Another name-idea to consider: **Automatically Run**

Apple includes a related feature for its "Playground" project-type in Xcode where a developer can freely type Swift and immediately see it executed.

I may be misunderstanding the nature of this discussion, and I hope I am not. Here is a screenshot of the Playground and the UI where the auto execution feature resides.

<img width="1728" alt="Screenshot 2023-06-02 at 8 49 28 PM" src="https://github.com/IndiePython/nodezator/assets/704698/68004ea3-a275-49a2-bfaf-74d9319029fd">

**Reply by KennedyRichard** [reply]
I'm not familiar with Playground, but if it automatically runs the code whenever it is changed, then I think the concept applies. Thank you for the suggestion.

I believe my first assumption was wrong, that is, that he meant a continuous/indefinite execution. For instance, maybe once per frame, etc.

Now, I believe what he wants is that Nodezator executes automatically in response to changes in the inputs. That is, whenever the user changes and input, he wants the graph to automatically run. This is even more simpler to achieve, but equally as relevant, as it is a common and useful workflow, and one I've wanted to include in Nodezator since a long time ago.

When it is time to implement this feature, I'll include "Automatically Run" in the list of possible names as well, though something like "Run on change" might be a bit more accurate, I believe.

**Comment by sefgit** [comment]
If we could have toolbar-buttons (with keyboard shortcuts) for these features to :
STEP, RUN, PAUSE, STOP, RESET 
nodezator's current graph.

**Reply by KennedyRichard** [reply]
Could you please provide more info/context about those actions you described? I mean, RUN is obvious but what would be the criteria for STEP, for instance? And for RESET as well. I don't recall seeing this kind of actions in other node editors. Would this be for a feature like an animation timeline?

**Comment by sefgit** [comment]
I understand this [Visualize Code](https://pythontutor.com/visualize.html#mode=edit) is "different", but it would be useful and powerful to "investigate" huge graph, especially if we could have "node inspector" (view internal states/variables) someday.


Another "not-related" tool which have node editor to run its simulation, with support of STEP/RUN simulation.
[Control-IO from factoryio.com](https://docs.factoryio.com/manual/drivers/control-io/)

**Reply by KennedyRichard** [reply]
Thank you for the further explanation and examples/links. Those are all interesting concepts/features. Certainly something to consider for the future of Nodezator.
